bin/propolis-server/src/lib/initializer.rs

```diff
@@ -610,66 +617,67 @@ impl<'a> MachineInitializer<'a> {
                     block::attach(vioblk.clone(), backend).unwrap();
                     chipset.pci_attach(bdf, vioblk.clone());
                     vioblk
                 }
                 DeviceInterface::Nvme => {
                     // Limit data transfers to 1MiB (2^8 * 4k) in size
                     let mdts = Some(8);
-                    let component = format!("nvme-{}", disk_name);
+                    let component = format!("nvme-{}", device_id);
                     let nvme = nvme::PciNvme::create(
-                        disk_name.to_owned(),
+                        device_id.to_string(),
                         mdts,
                         self.log.new(slog::o!("component" => component)),
                     );
-                    self.devices
-                        .insert(format!("pci-nvme-{bdf}"), nvme.clone());
+                    self.devices.insert(device_id.clone(), nvme.clone());
                     block::attach(nvme.clone(), backend).unwrap();
                     chipset.pci_attach(bdf, nvme.clone());
                     nvme
                 }
             };
 
-            if let Some((disk_id, backend)) = crucible {
-                let block_size = backend.block_size().await;
-                let prev = self.crucible_backends.insert(disk_id, backend);
-                if prev.is_some() {
-                    return Err(MachineInitError::DuplicateCrucibleBackendId(
-                        disk_id,
-                    ));
-                }
-
+            if let Some(crucible) = crucible {
+                let block_size = crucible.block_size().await;
                 let Some(block_size) = block_size else {
                     slog::error!(
                         self.log,
                         "Could not get Crucible backend block size, \
                         virtual disk metrics can't be reported for it";
-                        "disk_id" => %disk_id,
+                        "disk_id" => %backend_id,
                     );
                     continue;
                 };
 
-                // Register the block device as a metric producer, if we've been
-                // setup to do so. Note we currently only do this for the Crucible
-                // backend, in which case we have the disk ID.
-                if let Some(registry) = &self.producer_registry {
+                let prev =
+                    self.crucible_backends.insert(backend_id.clone(), crucible);
+                if prev.is_some() {
+                    return Err(MachineInitError::DuplicateCrucibleBackendId(
+                        backend_id.clone(),
+                    ));
+                }
+
+                // If metrics are enabled and this Crucible backend was
+                // identified with a UUID-format spec key, register this disk
+                // for metrics collection, using the key as the disk ID.
+                if let (Some(registry), SpecKey::Uuid(disk_id)) =
```

Comment:
Nexus will never tell `propolis-server` about a Crucible disk with non-UUID ID, right? if so, the `SpecKey::Name` path here, where there just won't be metrics collected for the disk, _shouldn't_ be reachable except if we've crafted a weird instance request by hand? if this shouldn't ever be reachable, it seems worth rejecting the spec /more precisely typing IDs we know _need_ to be UUID.

i do remember disk _names_ ending up in here somewhere, rather than IDs.. gonna go reread how that plumbing works unless you remember more immediately.

my real concern here is that this is where Nexus had been giving what are now `SpecKey::Name`, and that we'll lose Crucible metrics here. do you know if we have a test that exercises that?

Reply:
> Nexus will never tell propolis-server about a Crucible disk with non-UUID ID, right? if so, the SpecKey::Name path here, where there just won't be metrics collected for the disk, shouldn't be reachable except if we've crafted a weird instance request by hand? if this shouldn't ever be reachable, it seems worth rejecting the spec /more precisely typing IDs we know need to be UUID.

This makes sense. I'm a little leery of it because I think it makes the API a little weird: the ID for a given component can be either a UUID a Name, *unless* it's this one specific kind of component, in which case it has to be a UUID. This isn't horribly broken or anything, it's just a slightly sharp edge (and we'd also have to account for it in the non-Omicron tools/processes that let users attach ad-hoc Crucible volumes to Propolis servers).

We could address this in a few other ways:

1. `warn!` if the spec key isn't a UUID but there's a producer registry available
2. put a `#[cfg(feature = "omicron-build")]` guard in that enforces the "Crucible keys must be UUIDs" requirement
3. go back to calling `Volume::get_uuid` in this path to get the UUID to use for stats reporting, and `warn!` if it differs from the spec key

I kind of like option 3 now that I've written it up: it limits our regression risk (at least where metrics are concerned) but keeps the API flexible. WDYT?

> i do remember disk names ending up in here somewhere, rather than IDs.. gonna go reread how that plumbing works unless you remember more immediately.

IIRC the disk name wasn't used in this specific path, but it *was* used in VCR replacement, which worked like this:

- Take the VM objects lock
- Look up the `CrucibleStorageBackend` component in the instance spec
  - the key for this was derived from the disk name
- Look up the Crucible backend object in the VM's `crucible_backends` map
  - the key for this turns out to be the disk ID; the initialization code path used to read it from the Crucible backend's `get_uuid` function, which would read it out of the disk's volume construction request; Omicron guaranteed that this ID matched the disk ID (by creating the VCR that way)
- send the replace operation to Crucible
- on success, write the updated VCR back to the instance spec

> my real concern here is that this is where Nexus had been giving what are now SpecKey::Name, and that we'll lose Crucible metrics here. do you know if we have a test that exercises that?

We don't have an end-to-end test for disk metrics, at least AFAIK.

Reply:
yeah, after re-reading a bit and taking a second look, i'm also a fan of `Volume::get_uuid` here. i generally like the change to make `crucible` just `Option<Arc<...>>` though, so to double-check: you're thinking literally calling `crucible.get_uuid()` somewhere around this `if`, yeah? i think that should be pretty clean.

FWIW the `get_uuid()` call before was context'd and `?`, so it would bubble an error up rather than warn. that still seems like an appropriate error behavior here too.

more UUID tracing i haven't done: could `backend_id` be different from `crucible.get_uuid()`? i assume this is "technically possible but not really desired"

Reply:
Should be fixed in d6602c1.

> more UUID tracing i haven't done: could backend_id be different from crucible.get_uuid()? i assume this is "technically possible but not really desired"

That's correct, at least as I understand it:

- the serialized `VolumeConstructionRequest` that gets passed to the API is expected to be a `VolumeConstructionRequest::Volume` (which will have subvolumes of type `Region`)
- `Volume::get_uuid` returns the ID field in that top-level `Volume`
- when Nexus constructs the VCR for a new disk, it sets that ID to the disk ID

When I update Omicron to pick up this change, I expect that Nexus/sled-agent will use disk IDs as the spec keys for any `CrucibleStorageBackend`s they specify. That will preserve this relationship so long as Nexus keeps inserting the disk ID into its `VolumeConstructionRequest::Volume`s the way it does today.